File: src/controllers/backend/plot.ts
import { invoke } from '@tauri-apps/api/tauri'

// -------------------------------------------------------------------------------------------------

/* 
  Result for a single entry of an AFEC high level database tSNE map 
  struct PlotEntry in `src-tauri/plot`
 */

export interface PlotEntry {
  filename: string,
  x: number,
  y: number,
  categories: Array<string>,
  classes: Array<string> 
}

// -------------------------------------------------------------------------------------------------

/* 
  Calculate TSNE plot of an AFEC high level database tSNE map 
  fn create_plot `src-tauri/plot`
 */

export function createPlot(dbPath: string, perplexity: number, theta: number, epochs: number) {
  return invoke<PlotEntry[]>('create_plot', { dbPath, perplexity, theta, epochs });
}

File: src-tauri/src/plot.rs
mod database;
mod tsne;

// -------------------------------------------------------------------------------------------------

// Create a t-SNE plot for the given AFEC database with the given parameters
// This should be async, as plot generation takes a long while.
#[tauri::command(async)]
pub fn create_tsne_plot(
    db_path: String,
    theta: f32,
    perplexity: f32,
    epochs: usize,
) -> Result<Vec<tsne::PlotEntry>, String> {
    tsne::create_plot(db_path, theta, perplexity, epochs).map_err(|err| err.to_string())
}
